src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,435 @@
+			task->state = SOF_TASK_STATE_RUNNING;
+			must_run = true;
+		} else {
+			must_run = false;
+		}
+		spin_unlock_irq(&sch->lock, flags);
+
+		/*
+		 * While the lock is not held, the task can be cancelled, we
+		 * deal with it after the task completion.
+		 */
+		if (must_run) {
+			/*
+			 * task's .run() should only return either
+			 * SOF_TASK_STATE_COMPLETED or SOF_TASK_STATE_RESCHEDULE
+			 */
+			state = task_run(task);
+			if (state != SOF_TASK_STATE_COMPLETED &&
+			    state != SOF_TASK_STATE_RESCHEDULE) {
+				tr_err(&ll_tr,
+				       "zephyr_ll_run: invalid return state %u",
+				       state);
+				state = SOF_TASK_STATE_RESCHEDULE;
+			}
+		} else {
+			state = SOF_TASK_STATE_COMPLETED;
+		}
+
+		spin_lock_irq(&sch->lock, flags);
+
+		pdata = task->priv_data;
+
+		if (pdata->terminator) {
+			/*
+			 * zephyr_ll_task_free() is trying to free this task.
+			 * complete it and signal the semaphore to let the
+			 * function proceed
+			 */
+			zephyr_ll_task_done(sch, task);
+			k_sem_give(&pdata->sem);
+		} else if (state == SOF_TASK_STATE_COMPLETED) {
+			zephyr_ll_task_done(sch, task);
+		} else {
+			/*
+			 * task->state could've been changed to
+			 * SOF_TASK_STATE_CANCEL
+			 */
+			switch (task->state) {
+			case SOF_TASK_STATE_CANCEL:
+				zephyr_ll_task_done(sch, task);
+				break;
+			default:
+				/* reschedule */
+				list_item_del(&task->list);
+
+				zephyr_ll_task_insert_unlocked(sch, task);
+
+				task->start = sch->ll_domain->next_tick;
+			}
+		}
+
+		spin_unlock_irq(&sch->lock, flags);
+	}
+
+	notifier_event(sch, NOTIFIER_ID_LL_POST_RUN,
+		       NOTIFIER_TARGET_CORE_LOCAL, NULL, 0);
+}
+
+/* Called once for periodic tasks or multiple times for one-shot tasks */
+static int zephyr_ll_task_schedule(void *data, struct task *task, uint64_t start,
+				   uint64_t period)
+{
+	struct zephyr_ll *sch = data;
+	struct zephyr_ll_pdata *pdata;
+	struct task *task_iter;
+	struct list_item *list;
+	uint64_t delay = period ? period : start;
+	uint32_t flags;
+	int ret;
+
+	tr_info(&ll_tr, "task add %p %pU priority %d flags 0x%x", task, task->uid,
+		task->priority, task->flags);
+
+	irq_local_disable(flags);
+	if (!sch->n_tasks)
+		/*
+		 * If we're adding the first task to the scheduler, it is
+		 * possible that the scheduler has already run before on this
+		 * core and its thread has been aborted while holding the
+		 * spinlock, so we have to re-initialize it. Also note, that
+		 * the counter is decremented as the last operation in
+		 * zephyr_ll_task_done() before aborting the thread, so just
+		 * disabling local IRQs provides sufficient protection here.
+		 */
+		spinlock_init(&sch->lock);
+	irq_local_enable(flags);
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* check if task is already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock_irq(&sch->lock, flags);
+			return 0;
```

Comment:
Should we not complain here and  return an error ? Under what circumstances would be schedule a task twice ?

Reply:
Don't know, this is taken from the original ll-scheduler. We can try to remove it and see what breaks, but I'd rather do it later in a separate PR. Let me add a comment here.